drivers/i2c/i2c_cdns.c

```diff
@@ -0,0 +1,1067 @@
+		}
+	}
+
+	/* Set the output values */
+	*a = calc_div_a;
+	*b = calc_div_b;
+	*f = best_fscl;
+
+out:
+	return ret;
+}
+
+/**
+ * cdns_i2c_setclk - Set the serial clock rate for the I2C device
+ * @i2c_bus: Pointer to the I2C data structure
+ * @clk_in: I2C clock input frequency in Hz
+ *
+ * This function sets the serial clock rate for the I2C device by configuring
+ * the clock divisors in the device's control register. The device must be idle
+ * (i.e., not actively transferring data) before calling this function.
+ *
+ * The clock rate is determined by the following formula:
+ *	Fscl = Fpclk / (22 * (divisor_a + 1) * (divisor_b + 1))
+ * Where:
+ *	- Fscl is the desired I2C clock rate
+ *	- Fpclk is the input clock frequency
+ *	- divisor_a and divisor_b are the calculated divisors to achieve the desired clock rate
+ *
+ * The serial clock rate cannot exceed the input clock divided by 22. Common I2C clock
+ * rates are 100 KHz and 400 KHz.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+static int32_t cdns_i2c_setclk(struct cdns_i2c_data *i2c_bus, uint32_t clk_in)
+{
+	uint32_t div_a, div_b;
+	uint32_t ctrl_reg;
+	int32_t ret = 0;
+	uint32_t fscl = i2c_bus->i2c_clk;
+
+	/* Calculate the divider values */
+	ret = cdns_i2c_calc_divs(&fscl, clk_in, &div_a, &div_b);
+	if (ret != 0) {
+		goto out;
+	}
+
+	/* Set new divider values in the control register */
+	ctrl_reg = i2c_bus->ctrl_reg;
+	ctrl_reg &= ~(CDNS_I2C_CR_DIVA_MASK | CDNS_I2C_CR_DIVB_MASK);
+	ctrl_reg |= ((div_a << CDNS_I2C_CR_DIVA_SHIFT) |
+		     (div_b << CDNS_I2C_CR_DIVB_SHIFT));
+	i2c_bus->ctrl_reg = ctrl_reg;
+	cdns_i2c_writereg(i2c_bus, ctrl_reg, CDNS_I2C_CR_OFFSET);
+
+out:
+	return ret;
+}
+
+/**
+ * cdns_i2c_configure - Configures the I2C bus speed and initializes the I2C peripheral
+ * @dev: Pointer to the device structure representing the I2C bus
+ * @dev_config: Configuration value containing the desired I2C bus speed
+ *
+ * This function configures the I2C bus speed based on the value provided in
+ * @dev_config, which can be either I2C_SPEED_STANDARD (100 KHz) or I2C_SPEED_FAST (400 KHz).
+ * It then sets the appropriate clock for the I2C bus, verifies the clock is valid, and
+ * initializes the I2C peripheral. The configuration is saved to the device's data structure.
+ *
+ * Return: 0 on success, negative error value on failure
+ */
+static int32_t cdns_i2c_configure(const struct device *dev, uint32_t dev_config)
+{
+	int32_t ret = 0;
+	struct cdns_i2c_data *i2c_bus = (struct cdns_i2c_data *)dev->data;
+	uint32_t speed = I2C_SPEED_GET(dev_config);
+	uint32_t i2c_speed;
+
+	(void)k_mutex_lock(&i2c_bus->bus_mutex, K_FOREVER);
+
+	/* Check requested I2C Speed */
+	if (speed == I2C_SPEED_STANDARD) {
+		i2c_speed = CDNS_I2C_SPEED_STANDARD_HZ; /* 100 KHz */
+	} else if (speed == I2C_SPEED_FAST) {
+		i2c_speed = CDNS_I2C_SPEED_FAST_HZ; /* 400 KHz */
+	} else {
+		LOG_ERR("Unsupported I2C speed requested: %u", speed);
+		ret = -ERANGE;
+		goto out;
+	}
```

Comment:
I'm not sure why this needs to be done... This is also capable of fast mode plus which is currently impossible due to this 'check', that `cdns_i2c_setclk` function should handle whatever speed within that range

Reply:
We can remove it but it is untested feature. At least for me it is better to fail to indicate that feature is missing instead of letting things go wrong.

Reply:
I don't believe that is the case or useful, for example, this is calculated out by a formula, I wouldn't expect you to test every single integer and point on that divisor graph for a scl value... I would expect the RTL to already of been validated by that point

I think this should still be removed.

Reply:
It should also be updated within here so when `config_get` is called, it returns the "true" SCL value from the integer division it is doing which could be rounded down from the requested value